packages/apollo-engine-reporting/src/extension.ts

```diff
@@ -455,3 +408,122 @@ function defaultGenerateClientInfo({ request }: GraphQLRequestContext) {
+    } else {
+      return variables;
+    }
+  })();
+
+  // Note: we explicitly do *not* include the details.rawQuery field. The
+  // Engine web app currently does nothing with this other than store it in
+  // the database and offer it up via its GraphQL API, and sending it means
+  // that using calculateSignature to hide sensitive data in the query
+  // string is ineffective.
+  Object.keys(variablesToRecord).forEach(name => {
+    if (
+      !sendVariableValues ||
+      'sendNone' in sendVariableValues ||
+      ('exceptNames' in sendVariableValues &&
+        // We assume that most users will have only a few variables values to hide,
+        // or will just set {sendNone: true}; we can change this
+        // linear-time operation if it causes real performance issues.
+        sendVariableValues.exceptNames.includes(name)) ||
+      ('includeNames' in sendVariableValues &&
+        !sendVariableValues.includeNames.includes(name))
+    ) {
+      // Special case for private variables. Note that this is a different
+      // representation from a variable containing the empty string, as that
+      // will be sent as '""'.
+      details.variablesJson![name] = '';
+    } else {
+      try {
+        details.variablesJson![name] =
+          variablesToRecord[name] === undefined
+            ? ''
+            : JSON.stringify(variablesToRecord[name]);
+      } catch (e) {
+        // This probably means that the value contains a circular reference,
+        // causing `JSON.stringify()` to throw a TypeError:
+        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#Issue_with_JSON.stringify()_when_serializing_circular_references
+        details.variablesJson![name] = JSON.stringify(
+          '[Unable to convert value to JSON]',
+        );
+      }
+    }
+  });
+  return details;
+}
+
+// Helper for makeTraceDetails() to enforce that the keys of a modified 'variables'
+// matches that of the original 'variables'
+function cleanModifiedVariables(
+  originalKeys: Array<string>,
+  modifiedVariables: Record<string, any>,
+): Record<string, any> {
+  let cleanedVariables: Record<string, any> = {};
+  originalKeys.forEach(name => {
+    cleanedVariables[name] = modifiedVariables[name];
+  });
+  return cleanedVariables;
+}
+
+export function makeHTTPRequestHeaders(
+  http: Trace.IHTTP,
+  headers: Headers,
+  sendHeaders?: SendValuesBaseOptions,
+): void {
+  if (!sendHeaders || 'sendNone' in sendHeaders) {
+    return;
+  }
+  for (const [key, value] of headers) {
+    if (
+      ('exceptNames' in sendHeaders &&
+        // We assume that most users only have a few headers to hide, or will
+        // just set {sendNone: true} ; we can change this linear-time
+        // operation if it causes real performance issues.
+        sendHeaders.exceptNames.some(exceptHeader => {
+          // Headers are case-insensitive, and should be compared as such.
```

Comment:
I guess this isn't new code but I think we could calculate key.toLowerCase() once instead of on each element of the list.  The comment above is about trying to go from linear-time to constant-time by using a map or whatever and is overkill, but it's pretty trivial to just save key.toLowerCase once right after the `for (const [key` line...